src/storage/src/source/timestamper.rs

```diff
@@ -0,0 +1,398 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Timestamper using persistent collection
+use std::collections::{HashMap, VecDeque};
+use std::time::Duration;
+
+use anyhow::{bail, Context as _};
+use differential_dataflow::lattice::Lattice;
+use futures::StreamExt;
+use mz_dataflow_types::client::controller::storage::CollectionMetadata;
+use timely::progress::{Antichain, Timestamp as _};
+use timely::PartialOrder;
+
+use mz_dataflow_types::sources::MzOffset;
+use mz_expr::PartitionId;
+use mz_ore::now::NowFn;
+use mz_persist_client::read::{ListenEvent, ListenStream, ReadHandle};
+use mz_persist_client::write::WriteHandle;
+use mz_persist_client::Upper;
+use mz_repr::{Diff, Timestamp};
+use tracing::{error, info};
+
+pub struct CreateSourceTimestamper {
```

Comment:
Can we add doc comments on as many thing as possible in this file? Its logic is very subtle and we should make extra sure we understand what it's doing and how it should behave.

The thing that is the most difficult to me to understand is how the frontier returned by the `timestamp_offsets` method is calculated and what is its relation to the user of this operator. But it'd be good to document all the fields of this struct too and their intended purpose.

Also, I'd advocate for a different name that corresponds better to our design docs (which refer to the process implemented here as reclocking). My suggestion would be to move this to `reclock.rs` and name this `RelockOperator` but I'm happy with any name/structure that makes the connection to our design docs clear.